### PR TITLE
CLI: orchestration commands

### LINKED ISSUE #239: CLI: upgrade kyma command
**Description**

Implement the `Run()` method of `kcp upgrade kyma` command to upgrade or reconfigure Kyma on one or more Kyma runtimes

The command triggers the `POST /upgrade/kyma` API of KEB with the supplied parameters to start a new Kyma Upgrade orchestration. The output is the orchestration ID returned as part of the response of the API request. 

**Acceptance Criteria**

- Implement orchestration client with the `UpgradeKyma()` method under https://github.com/kyma-project/control-plane/tree/master/components/kyma-environment-broker/common/orchestration
- Implement the `Run()` method using `orchestration.Client.UpgradeKyma()`

### LINKED ISSUE #240: CLI: orchestrations command
**Description**
Implement the `Run()` method of the kcp orchestrations command to display orchestrations.

The commands has two modes:

  1. Without specifying an orchestration id as an argument, it will list all orchestrations, or orchestrations matching the --state if supplied.
  2. When specifying an orchestration id as an argument, it will display details about the specific orchestration.

**List orchestrations**
The command triggers GET `/orchestrations` KEB API. Filtering on state should be supported (also planned to be implemented on KEB server side in 1.17)

The output should display the following attributes:

- Orchestration ID
- Type (Currently only "Kyma Upgrade")
- Created At
- State

The output format can be 

- Table (default)
- JSON
- YAML

Filtering on State should be supported (both by server and CLI) using the --state <succeeded | inprogress | failed> command line argument.

**Get orchestration mode**
Triggers GET `/orchestrations/{orchestration_id}` KEB API. 

Usage: `skr orchestration get <ORCHESTRATION ID> [–-operation <OPERATION ID>]`.

Without operation option the output displays the following attributes:

Orchestration ID
Type (Currently only "Kyma Upgrade")
Created At
State
Parameters
Runtime Operations

The output format can be table (default), JSON, YAML

Table output example
```
Orchestration ID   : xxxx-xxx-xxxxxxx-xxxxx
Type               : Kyma Upgrade
Created At         : 2020-09-09-15:00:00Z
State              : Succeeded
Parameters         : <JSON blob in one-line format ?>
Runtime Operations : 
OPERATION ID             SHOOT     GLOBALACCOUNT ID         SUBACCOUNT ID            RUNTIME ID               STATE
iiiiii-iiii-iiii-iiiiii  abcdefg   yyyyyy-yyyy-yyyy-yyyyyy  zzzzzz-zzzz-zzzz-zzzzzz  rrrrrr-rrrr-rrrr-rrrrrr  Succeeded
[...]
[...]
```
**Acceptance Criteria**

- Implement orchestration client with the `ListOrchestrations()` and `GetOrchestration(id)` methods under https://github.com/kyma-project/control-plane/tree/master/components/kyma-environment-broker/common/orchestration
- Implement the `Run()` method using `orchestration.Client` methods

### PR DESCRIPTION
**Pull request by ebensom**
**Description**

CLI: orchestration commands

Changes proposed in this pull request:

- Changed auth model of KEB `/orchestrations` and `/upgrade` endpoints to be based on OIDC JWT verification
- Implemented `kcp orchestrations` and `kcp upgrade kyma` commands
- Fixed immediate and maintenance windows schedules of upgrade operations

**Related issue(s)**
Resolves #239 
Resolves #240

### CODE REVIEW

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 1]
```diff
@@ -59,7 +59,7 @@ spec:
 EOF
 ```
 
-> **NOTE:** The valid scopes are `broker:write`, `broker-upgrade:read`, `broker-upgrade:write`, and `cld:read`.
+> **NOTE:** The valid scopes are `broker:write`, and `cld:read`.
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-05-authorization.md:62]:
```suggestion
> **NOTE:** The valid scopes are `broker:write` and `cld:read`.
```

File: docs/kyma-environment-broker/08-05-orchestrate-kyma-upgrade.md [hunk 2]
```diff
@@ -13,10 +13,10 @@ This tutorial shows how to upgrade Kyma Runtime using Kyma Environment Broker.
 
 ## Steps
 
-1. [Get the access token](#details-authorization). Export this variable based on the token you got from the OAuth client:
+1. [Get the OIDC ID token in JWT format](#details-orchestration). Export this variable based on the token you got from the OIDC client:
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/08-05-orchestrate-kyma-upgrade.md:16]:
```suggestion
1. [Get the OIDC ID token in the JWT format](#details-orchestration). Export this variable based on the token you got from the OIDC client:
```

File: docs/kyma-environment-broker/03-10-orchestration.md [hunk 3]
```diff
@@ -5,9 +5,9 @@ type: Details
...
 
-If Kyma Environment Broker is restarted, it reprocesses the orchestration with the `IN PROGRESS` state. 
+If Kyma Environment Broker is restarted, it reprocesses the orchestration with the `IN PROGRESS` state.
 
->**NOTE:** You need a token with the `broker-upgrade:write` authorization scope to create an orchestration, and a token with the `broker-upgrade:read` scope to fetch the orchestrations.
+>**NOTE:** You need an OIDC ID token in JWT format, issued by a (configurable) OIDC provider which is trusted by Kyma Environment Broker. The `groups` claim must be present in the token, and furthermore the user must belong to the configurable admin group (`runtimeAdmin` by default) to create an orchestration, and must belong to the configurable operator group (`runtimeOperator` by default) to fetch the orchestrations.
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-10-orchestration.md:10]:
```suggestion
>**NOTE:** You need an OIDC ID token in the JWT format issued by a (configurable) OIDC provider which is trusted by Kyma Environment Broker. The `groups` claim must be present in the token, and the user must belong to the configurable admin group (`runtimeAdmin` by default) to create an orchestration. To fetch the orchestrations, the user must belong to the configurable operator group (`runtimeOperator` by default).
```

File: docs/cli/commands/kcp_upgrade_kyma.md [hunk 4]
```diff
@@ -27,14 +27,15 @@ kcp upgrade kyma --target {TARGET SPEC} ... [--target-exclude {TARGET SPEC} ...]
       --dry-run                      Perform the orchestration without executing the actual upgrage operations for the Runtimes. The details can be obtained using the "kcp orchestrations" command.
       --parallel-workers int         Number of parallel workers to use in parallel orchestration strategy. By default the amount of workers will be auto-selected on control plane server side.
       --schedule string              Orchestration schedule to use. Possible values: "immediate", "maintenancewindow". By default the schedule will be auto-selected on control plane server side.
-      --strategy string              Orchestration strategy to use. Currently the only supported strategy is parallel. (default "parallel")
+      --strategy string              Orchestration strategy to use. The possible values are: parallel. (default "parallel")
```
Comment (klaudiagrz) [inline docs/cli/commands/kcp_upgrade_kyma.md:30]:
```suggestion
      --strategy string              Orchestration strategy to use. The default value is "parallel". 
```

Comment (klaudiagrz) [reply docs/cli/commands/kcp_upgrade_kyma.md:30]: ... or `The only possible value is "parallel".

**Review by klaudiagrz — DISMISSED** [review]
LGTM 🦊

File: components/kyma-environment-broker/cmd/cli/command/options.go [hunk 5]
```diff
@@ -23,10 +23,17 @@ const (
...
+	subaccountTarget = "subaccount"
+	runtimeIDTarget  = "runtime-id"
+	regionTarget     = "region"
+	planTarget       = "plan"
+)
+
+const (
+	azurePlan     = "azure"
+	azureLitePlan = "azure_lite"
+	trialPlan     = "trial"
```
Comment (polskikiel) [inline components/kyma-environment-broker/cmd/cli/command/options.go:36]: don't we want to support all the plans available under `internal/broker/plans.go`?

File: components/kyma-environment-broker/cmd/cli/command/orchestration.go [hunk 6]
```diff
@@ -1,23 +1,124 @@
...
+Orchestration ID   : {{.OrchestrationID}}
+Global Account ID  : {{.GlobalAccountID}}
+Subaccount ID      : {{.SubAccountID}}
+Runtime ID         : {{.RuntimeID}}
+Shoot Name         : {{.ShootName}}
+Service Plan       : {{.ServicePlanName}}
+Maintenance Window : {{.MaintenanceWindowBegin}} - {{.MaintenanceWindowEnd}}
+State              : {{.State}}
+Description        : {{.Description}}
+Kyma Version       : {{.KymaConfig.Version}}
```
Comment (polskikiel) [inline components/kyma-environment-broker/cmd/cli/command/orchestration.go:119]: I think we can expose here also the k8s version

File: docs/kyma-environment-broker/08-05-orchestrate-kyma-upgrade.md [hunk 7]
```diff
@@ -13,10 +13,10 @@ This tutorial shows how to upgrade Kyma Runtime using Kyma Environment Broker.
 
 ## Steps
 
-1. [Get the access token](#details-authorization). Export this variable based on the token you got from the OAuth client:
+1. [Get the OIDC ID token in the JWT format](#details-orchestration). Export this variable based on the token you got from the OIDC client:
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/08-05-orchestrate-kyma-upgrade.md:16]:
```suggestion
1. [Get the OIDC ID token in the JWT format](#details-authorization). Export this variable based on the token you got from the OIDC client:
```

File: docs/kyma-environment-broker/08-06-orchestration-status.md [hunk 8]
```diff
@@ -17,7 +17,7 @@ This tutorial shows how to check the orchestration status. Using this API, you c
    export ORCHESTRATION_ID={OBTAINED_ORCHESTRATION_ID}
    ```
 
-2. Make a call to the Kyma Environment Broker with a proper **Authorization** [request header](#details-authorization) to verify that the orchestration succeeded.
+2. Make a call to the Kyma Environment Broker with a proper **Authorization** [request header](#details-orchestration) to verify that the orchestration succeeded.
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/08-06-orchestration-status.md:20]:
```suggestion
2. Make a call to the Kyma Environment Broker with a proper **Authorization** [request header](#details-authorization) to verify that the orchestration succeeded.
```

File: docs/kyma-environment-broker/08-06-orchestration-status.md [hunk 9]
```diff
@@ -53,7 +53,7 @@ This tutorial shows how to check the orchestration status. Using this API, you c
 
 ## Fetch all orchestrations status
 
-Make a call to the Kyma Environment Broker with a proper **Authorization** [request header](#details-authorization) to verify that the orchestration succeeded.
+Make a call to the Kyma Environment Broker with a proper **Authorization** [request header](#details-orchestration) to verify that the orchestration succeeded.
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/08-06-orchestration-status.md:56]:
```suggestion
Make a call to the Kyma Environment Broker with a proper **Authorization** [request header](#details-authorization) to verify that the orchestration succeeded.
```

File: docs/kyma-environment-broker/08-06-orchestration-status.md [hunk 10]
```diff
@@ -98,14 +98,14 @@ A successful call returns the list of all orchestrations:
    export ORCHESTRATION_ID={OBTAINED_ORCHESTRATION_ID}
    ```
 
-2. Make a call to the Kyma Environment Broker with a proper **Authorization** [request header](#details-authorization) to fetch the list of the upgrade operations for a given orchestration.
+2. Make a call to the Kyma Environment Broker with a proper **Authorization** [request header](#details-orchestration) to fetch the list of the upgrade operations for a given orchestration.
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/08-06-orchestration-status.md:101]:
```suggestion
2. Make a call to the Kyma Environment Broker with a proper **Authorization** [request header](#details-authorization) to fetch the list of the upgrade operations for a given orchestration.
```

File: docs/kyma-environment-broker/08-06-orchestration-status.md [hunk 11]
```diff
@@ -150,14 +150,14 @@ A successful call returns the list of all orchestrations:
    export OPERATION_ID={SCHEDULED_OPERATION_ID}
    ```
 
-2. Make a call to the Kyma Environment Broker with a proper **Authorization** [request header](#details-authorization) to fetch the details for given operation.
+2. Make a call to the Kyma Environment Broker with a proper **Authorization** [request header](#details-orchestration) to fetch the details for given operation.
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/08-06-orchestration-status.md:153]:
```suggestion
2. Make a call to the Kyma Environment Broker with a proper **Authorization** [request header](#details-authorization) to fetch the details for given operation.
```

**Review by klaudiagrz — COMMENTED** [review]
As @polskikiel  pointed out, these links should lead to the document re authorization, since it describes how to get the token and how to specify the request header.

File: docs/kyma-environment-broker/08-05-orchestrate-kyma-upgrade.md (hunk 7 above)
Comment (ebensom) [reply docs/kyma-environment-broker/08-05-orchestrate-kyma-upgrade.md:16]: I changed the authorization model of the orchestration endpoints to not be based or ORY (Oauth2), but be based on OIDC JWT verification, where the OIDC ID token is issued by an external OIDC provider. I also changed these links to not point to the Oauth2 authorization guide any more, but to the general orchestration guide where the requirements towards the  OIDC ID token arte mentioned.

**Review by klaudiagrz — DISMISSED** [review]

**Review by klaudiagrz — APPROVED** [review]

**Review by polskikiel — APPROVED** [review]

**Pull request merged**